File: kernel/include/kernel/serial.c
// https://wiki.osdev.org/Serial_Ports
#define PORT 0x03F8 // COM1

#include <kernel/io.h>

int init_serial() {
    outb(PORT + 1, 0x00); // Disable all interrupts
    outb(PORT + 3, 0x80); // Enable DLAB (set baud rate divisor)
    outb(PORT + 0, 0x03); // Set divisor to 3 (lo byte) 38400 baud
    outb(PORT + 1, 0x00); //                  (hi byte)
    outb(PORT + 3, 0x03); // 8 bits, no parity, one stop bit
    outb(PORT + 2, 0xC7); // Enable FIFO, clear them, with 14-byte threshold
    outb(PORT + 4, 0x0B); // IRQs enabled, RTS/DSR set
    outb(PORT + 4, 0x1E); // Set in loopback mode, test the serial chip
    outb(PORT + 0, 0xAE); // Test serial chip (send byte 0xAE and check if serial returns same byte)

    // Check if serial is faulty (i.e: not same byte as sent)
    if(inb(PORT + 0) != 0xAE) {
        return 1;
    }

    // If serial is not faulty set it in normal operation mode
    // (not-loopback with IRQs enabled and OUT#1 and OUT#2 bits enabled)
    outb(PORT + 4, 0x0F);
    return 0;
}

int is_transmit_empty() {
    return inb(PORT + 5) & 0x20;
}

void write_serial(char c) {
    while (is_transmit_empty() == 0);
    outb(PORT, c);
}

void write_serial_string(char *str) {
    while (*str != '\0') {
        write_serial(*str);
        str++;
    }
}

int serial_received() {
    return inb(PORT + 5) & 1;
}

char read_serial() {
    while (serial_received() == 0);
    return inb(PORT);
}

File: kernel/src/serial.zig
const io = @import("io.zig");

const COM1_PORT = 0x03F8;

pub fn initialize() void {
    io.outb(COM1_PORT + 1, 0x00); // disable all interrupts
    io.outb(COM1_PORT + 3, 0x80); // enable DLAB (set baud rate divisor)
    io.outb(COM1_PORT + 0, 0x03); // set divisor to 3 (lo byte) 38400 baud
    io.outb(COM1_PORT + 1, 0x00); //                  (hi byte)
    io.outb(COM1_PORT + 3, 0x03); // 8 bits, no parity, one stop bit
    io.outb(COM1_PORT + 2, 0xC7); // enable FIFO, clear them, with 14-byte threshold
    io.outb(COM1_PORT + 4, 0x0B); // IRQs enabled, RTS/DSR set
    io.outb(COM1_PORT + 4, 0x1E); // set in loopback mode, test the serial chip
    io.outb(COM1_PORT + 0, 0xAE); // test serial chip (send byte 0xAE and check if serial returns same byte)

    if (io.inb(COM1_PORT) != 0xAE)
        return;

    io.outb(COM1_PORT + 4, 0x0F);
}

fn is_transmit_empty() u8 {
    return io.inb(COM1_PORT + 5) & 0x20;
}

pub fn write(character: u8) void {
    while (is_transmit_empty() == 0) {}
    io.outb(COM1_PORT, character);
}

pub fn write_string(string: []const u8) void {
    for (string) |char|
        write(char);
}
